File: easy/same_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <iostream>
#include <vector>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        vector<int> pVec;
        vector<int> qVec;
        
        storeValues(pVec, p);
        storeValues(qVec, q);
        
        if (pVec.size() != qVec.size()) {
            return false;
        }
        
        for (int i = 0; i < pVec.size(); i++) {
            if (pVec[i] != qVec[i]) {
                return false;
            }
        }
        
        return true;
    }
    
    void storeValues(vector<int> &vec, TreeNode* root) {
        if (root == nullptr) {
            vec.push_back(-1);
            return;
        }
        
        vec.push_back(root->val);
        storeValues(vec, root->left);
        storeValues(vec, root->right);
        
        
    }
};

File: blind-75/easy/same_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # traverse through a tree

        if not p and not q:
            return True
        
        if p and q and p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

        # def helper(root, lst):
        #     if not root:
        #         return lst
            
        #     lst.append(root.val)

        #     helper(root.left, lst)
        #     helper(root.right, lst)

        # lst_p = helper(p, [])
        # lst_q = helper(q, [])

        # if lst_p == lst_q:
        #     return True

        # return False
        
